### ISSUE TITLE
Bar can grow beyond its allocated strut size

### ISSUE THREAD

**Issue by PierreR** [taffybar/taffybar#270]
I have just switched to taffybar-1.0.1. Everything is great so far except for the struts height. I guess 2 pictures will describe the problem more clearly:
![taffybar_1](https://user-images.githubusercontent.com/93115/38690208-d966790a-3e7d-11e8-9fe0-4d7c9427de7f.PNG)
![taffybar_2](https://user-images.githubusercontent.com/93115/38690209-d98570ee-3e7d-11e8-838c-f0e84e963290.PNG)

I am using a faily standard configuration (the xmonad one has all the bit from the provided example).

Any pointer where to look at to mitigate the problem ?

Thanks for your help

**Comment by IvanMalison** [comment]
@PierreR My guess is that the bar height that you've set is not quite large enough to accomodate the widgets that you have, and that the widgets are growing the strut window, without actually growing the strut height.

Can you post the exact config you are using?

**Comment by PierreR** [comment]
The only change from the example configuration is a font change. I will check again with the vanilla configuration from the this repo.

I have realized that the missing space is actually the one from the borderWidth property of [`WorkspaceHUDConfig`](https://github.com/travitch/taffybar/blob/master/src/System/Taffybar/WorkspaceHUD.hs#L279). So if I set it to 0, my problem vanished.

**Comment by PierreR** [comment]
I actually prefer it with a borderWith of 0, so in a way this is not an issue for me anymore ;-)

Shall I close it ?

**Comment by LeifW** [comment]
I have a similar issue when upgrading to the GTK3 version; I had to set `barHeight = 35` in `defaultTaffybarConfig` to keep the bottom of graph widgets from being cut off.

**Comment by IvanMalison** [comment]
> Maybe I need to account for that in my xmonad.hs config file ?

You mean taffybar.hs, right? You just need to make the barSize large.

> Shall I close it ?

It seems that a lot of people are a bit confused on this point, so I think it is probably worth trying to come up with a way to prevent this from happening.

Basically, the issue is that the bar can grow larger than the specified barSize, which can lead to a mismatch between the amount of space reserved for the bar with strut properties and the actual bar size.

I think the solution is to change the strut property when the barSize changes automatically.

It may even make sense to do away with barSize completely, and just let it be determined by what widgets get added, since I'm assuming that is the behavior that most people would want. I guess the problem with that is that you then have to manually make sure that all your widget sizes agree. It would be nice to be able to say: I want all the widgets to size themselves so that they come out to have x height, and have them automatically figure out how to make themselves that size. The reason that something like that might be tricky is that css, and a whole bunch of other stuff can affect size, so its hard to write widgets that can get their size to be something specific.

Those are my thoughts on the topic... I'm open to suggestions.

**Comment by PierreR** [comment]
> You mean taffybar.hs, right? You just need to make the barSize large.

By changing this property ? https://github.com/travitch/taffybar/blob/master/src/System/Taffybar.hs#L264


>  I think the solution is to change the strut property when the barSize changes automatically.

Still a bit confused about this ... How do we change the strut property ?

**Comment by IvanMalison** [comment]
@PierreR Yes, increasing that number should make things better for now.

> Still a bit confused about this ... How do we change the strut property ?

The strut property is currently set with a gdk_change_property bindings. In the future, and actually in this branch that you can (and should) start using now:

https://github.com/travitch/taffybar/tree/use_gtk-strut

this is actually handled in another libarary (https://github.com/IvanMalison/gtk-strut).

**Comment by IvanMalison** [comment]
@travitch or anyone else who has an opinion:

I think my plan to deal with this is just to have the strut allocation adapt to the bar size if the bar does end up expanding.

The other option would be to try to aggresively prevent expansion of the bar.

**Comment by travitch** [comment]
I think that is probably the easiest approach for now, especially if gtk doesn’t
give us tools for constraining things properly.  It might be nice longer term to
cook up something, but it would be invasive if we have to manage it ourselves.

**Comment by PierreR** [comment]
Deprecated issue. Only relevant for taffybar-1.x

**Comment by IvanMalison** [comment]
@PierreR  I don't think that is true.

**Comment by illis** [comment]
I'm having the issue with taffybar 2.1.1 -> as described in gitter

**Comment by illis** [comment]
Sorry that should be 2.1.2

**Comment by Magicloud** [comment]
Similar issue, not sure root cause is the same.

Latest source from github. Font size in css is 15pt, barHeight in taffybar.hs is 24.

I have two screen attached, so I can compare. The actual barHeight is larger than strut avoid size. About 1/4 is covered by application windows.

**Comment by IvanMalison** [comment]
you need to either increase barHeight, or decrease barSize...

**Comment by IvanMalison** [comment]
pt != px

**Comment by IvanMalison** [comment]
@Magicloud Well, then some other element is responsible for making the bar larger than the height that you have specified. Perhaps padding in your CSS? Could also be images. If you paste an image it might be easier for me ot see.


I meant that you need to increase barHeight. Not sure why I said decrease.

**Comment by Magicloud** [comment]
@IvanMalison Thanks. Finally I got time to confirm the issue is in the default .css. Some padding-s.

**Comment by qrilka** [comment]
@PierreR could you show your config (and probably css)? I have tried the default ones and appearance of selected workspace looks very bad, nothing like what's on your screenshot

**Comment by PierreR** [comment]
@qrilka well the css was working with taffybar-1.x at the time but since then many changes have been made in taffybar so it won't be of much use.

@IvanMalison It would be nice to have a handful of working css within this repo and some screenshot to improve the overall user experience. I would share mine when I am ready to use  taffybar 3.x.

**Comment by qrilka** [comment]
@IvanMalison I've used that CSS (made workspace label font smaller and decreased padding) and workspace selection doesn't look great - https://imgur.com/a/8v7c65Q
Is it expected to be just some white bar under an icon? I guess I need to fiddle with CSS e.g. to change workspace label color/background?

**Comment by PierreR** [comment]
@IvanMalison I might be missing something but on `nixos` this is what you get using the current default css (just removing the battery and the mpris2New widget.
![taffy](https://user-images.githubusercontent.com/93115/44862180-6da02700-ac7a-11e8-9343-f708e6f73a3a.PNG)
I know it can be changed to look better. I just want to express that the default provided css looks really bad.

**Comment by IvanMalison** [comment]
@PierreR The default taffybar configuration is meant to be used with a compositor that allows transparency. It seems like you aren't using one.

What exactly "looks bad" about it to you?

**Comment by qrilka** [comment]
@IvanMalison I was using version older version `0.4.6` with no problems on bare bones xmonad (without any special tweaks) so now some extra care is required? Any hints how this could be done? Is this reliance on transparency documented in any way?

**Comment by IvanMalison** [comment]
@qrilka I can change the default css so that it works without transparency if you tell me what you think sensible defaults would be. What exactly is it that you feel does not look good.

**Comment by qrilka** [comment]
@IvanMalison I think the main inconvenience for me is with workspace selection - on my notebook with old taffybar it shows selected workspace label in square brackets and with yellow font.
Also there seem to be a couple of other minor issues but those are not probably about css so I will file them as separate tickets.

**Comment by IvanMalison** [comment]
@qrilka Yes, that is not so much a cosmetic change as a change to the way the workspace widget works completely. I personally do not like the brackets strategy (I think a color or a box surrounding the workspace looks much nicer, which is why that is the default approach now). You can restore the old behavior of square brackets using this property:

https://github.com/taffybar/taffybar/blob/09c600cdc1332619087a90c3958d5d1fd68a6067/src/System/Taffybar/Widget/Workspaces.hs#L158

You can set background color of the active workspace using this property:

https://github.com/taffybar/taffybar/blob/09c600cdc1332619087a90c3958d5d1fd68a6067/taffybar.css#L56

You can set the label color by setting the foreground color on that same class.

**Comment by qrilka** [comment]
@PierreR any pointers how that gets done for those who are not aware?

**Comment by dpnova** [comment]
I have a similar issue I think, when i have two screens of different sizes I get the bar from one, over the other 

![Image](https://github.com/user-attachments/assets/34d5f231-02e0-4900-9ffa-2e1d50b12ad0)
